mcs/class/corlib/System/Array.cs

```diff
@@ -204,16 +208,31 @@ internal void InternalArray__set_Item<T> (int index, T item)
 				oarray [index] = (object)item;
 				return;
 			}
+			// Do not change this to call SetGenericValue_icall directly, due to special casing in the runtime.
 			SetGenericValueImpl (index, ref item);
 		}
 
 		// CAUTION! No bounds checking!
 		[MethodImplAttribute (MethodImplOptions.InternalCall)]
-		internal extern void GetGenericValueImpl<T> (int pos, out T value);
+		extern static void GetGenericValue_icall<T> (ref Array self, int pos, out T value);
 
 		// CAUTION! No bounds checking!
 		[MethodImplAttribute (MethodImplOptions.InternalCall)]
-		internal extern void SetGenericValueImpl<T> (int pos, ref T value);
+		extern static void SetGenericValue_icall<T> (ref Array self, int pos, ref T value);
+
+		// This is a special case in the runtime.
+		internal void GetGenericValueImpl<T> (int pos, out T value)
+		{
+			var self = this;
+			GetGenericValue_icall (ref self, pos, out value);
+		}
+
+		// This is a special case in the runtime.
+		internal void SetGenericValueImpl<T> (int pos, ref T value)
```

Comment:
Protect from inlining?

Reply:
Good point, I hadn't thought about it or verified it, that intrinsification should dominate inlining, but, um 1. I think it is fairly clearly ordered that way in the JIT 2. All the other functions, many of which, I think also have IL.
It is worth I guess some asserts, however, that will require sharing intrinsic-identification logic.

Reply:
> I think it is fairly clearly ordered that way in the JIT 
Ah ok, I wasn't sure.

> It is worth I guess some asserts

I wasn't necessarily suggesting that.  More just wondering whether you've seen evidence that  `emit_array_generic_access` is getting called after your changes.

Reply:
I admit, I didn't look, just assumed.
Except, I do know the signature change/index is needed.
Things failed without it.
I should do some verbose or aot dumps.